lib/util.js

```diff
@@ -738,7 +738,9 @@ function formatSet(ctx, value, recurseTimes, visibleKeys, keys) {
   var output = [];
   value.forEach(function(v) {
     var nextRecurseTimes = recurseTimes === null ? null : recurseTimes - 1;
-    var str = formatValue(ctx, v, nextRecurseTimes);
+    var str = ctx.seen.includes(v) ?
+      ctx.stylize('[Circular]', 'special') :
+      formatValue(ctx, v, nextRecurseTimes);
```

Comment:
@not-an-aardvark @BridgeAR @aqrln This is what I was having in mind:

<details><summary>diff in the fold</summary>

```diff
diff --git a/lib/util.js b/lib/util.js
index a9f267bd416d..9b8e8395ffac 100644
--- a/lib/util.js
+++ b/lib/util.js
@@ -216,7 +216,7 @@ function debuglog(set) {
 function inspect(obj, opts) {
   // default options
   var ctx = {
-    seen: [],
+    seen: new Set(),
     stylize: stylizeNoColor
   };
   // legacy...
@@ -613,11 +613,16 @@ function formatValue(ctx, value, recurseTimes) {
     }
   }
 
-  ctx.seen.push(value);
+  var output;
+  if (ctx.seen.has(value)) {
+    output = ctx.stylize('[Circular]', 'special');
+  } else {
+    ctx.seen.push(value);
 
-  var output = formatter(ctx, value, recurseTimes, visibleKeys, keys);
+    output = formatter(ctx, value, recurseTimes, visibleKeys, keys);
 
-  ctx.seen.pop();
+    ctx.seen.pop();
+  }
 
   return reduceToSingleString(output, base, braces, ctx.breakLength);
 }
@@ -822,21 +827,17 @@ function formatProperty(ctx, value, recurseTimes, visibleKeys, key, array) {
     }
   }
   if (!str) {
-    if (ctx.seen.indexOf(desc.value) < 0) {
-      if (recurseTimes === null) {
-        str = formatValue(ctx, desc.value, null);
+    if (recurseTimes === null) {
+      str = formatValue(ctx, desc.value, null);
+    } else {
+      str = formatValue(ctx, desc.value, recurseTimes - 1);
+    }
+    if (str.indexOf('\n') > -1) {
+      if (array) {
+        str = str.replace(/\n/g, '\n  ');
       } else {
-        str = formatValue(ctx, desc.value, recurseTimes - 1);
+        str = str.replace(/^|\n/g, '\n   ');
       }
-      if (str.indexOf('\n') > -1) {
-        if (array) {
-          str = str.replace(/\n/g, '\n  ');
-        } else {
-          str = str.replace(/^|\n/g, '\n   ');
-        }
-      }
-    } else {
-      str = ctx.stylize('[Circular]', 'special');
     }
   }
   if (name === undefined) {
```
</details>

WDYT?